### PR TITLE
Switch to a whitelist config that specifies types and fields.

### PR DESCRIPTION
**Pull request by minasmart**
Previous behaviour would only  whitelist types and strip types and fields of types not in the list.

In this version, we capture types and the fields used on that type, which should let us strip even more from the compiled schema

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -9,21 +9,23 @@ function yieldTypes(schema) {
...
 
-function filterTypes(whitelist) {
+function filterTypes(whitelistConfig) {
   return function(types) {
-    if (!whitelist) {
+    if (!whitelistConfig) {
       return types;
     }
 
+    const whitelistTypes = Object.getOwnPropertyNames(whitelistConfig);
```
Comment (swalkinshaw) [inline src/index.js]: Probably doesn't matter for this, but shouldn't `Object.keys` by used here?

File: src/cli.js [hunk 2]
```diff
@@ -17,8 +17,15 @@ function runCli() {
     process.exit(0);
   }
 
+  let whitelistConfig;
+
+  if (args.whitelistConfig) {
+    whitelistConfig = JSON.parse(fs.readFileSync(args.whitelistConfig));
+  }
+
+
```
Comment (lyninx) [inline src/cli.js]: ✂️

File: src/index.js (hunk 1 above)
Comment (minasmart) [inline src/index.js]: I learned a thing! `getOwnPropertyNames` ensures that you're just getting the properties you define on a thing. If something changes the prototype of Object and adds props, this makes sure they aren't captured here. It's only properties declared directly on the instance. It's a subtle difference, but something that seems safer to do in a library?

thoughts?

Comment (swalkinshaw) [inline src/index.js]: `keys` should act the same. ie: own properties only. Only difference is it doesn't return non-enumerable properties. `for .. in ` is the method that also has properties from the prototype chain (according to MDN).

Comment (minasmart) [inline src/index.js]: oh cool! I think I must have read something not entirely right at some point. Will update.

**Comment by minasmart** [comment]
@swalkinshaw I'm not sure  why this never got merged (please remind me if you remember) but changes made. ready to merge. let me know if you feel it's ready.

**Comment by swalkinshaw** [comment]
Think it just got lost in the shuffle 👍

**Pull request merged**